### PR TITLE
`azurerm_role_definition` - swap to `go-azure-sdk`

### PR DESCRIPTION
**Pull request by ziyeqf**
swap from `azure-sdk-for-go` to `go-azure-sdk`, keep the same API version.
In next PR, it will be upgraded to `2022-05-01-preview`,  for feature request.

And one more question is about the id of `azurerm_role_definition`, I made on #23679 (but removed from it):

The customized resource Id was invovled in #6107,  at that time, the resource id set to state file came from Azure, and in some case the scope was not included in the resource ID, e.g: role definition with scope of `management group`.
 
But for `go-azure-sdk`, the id is `ScopedRoledefinitionId` which always contain a scope, maybe we do not need to use the customized resource Id?  If the answer is yes, I will be glad to open another PR to change the resource id.

Test
---
```
terraform-provider-azurerm tengzh/rbac/sdk_swap ≡3s 
❯ tftest authorization TestAccRoleDefinition     
=== RUN   TestAccRoleDefinitionDataSource_basic
=== PAUSE TestAccRoleDefinitionDataSource_basic
=== RUN   TestAccRoleDefinitionDataSource_basicByName
=== PAUSE TestAccRoleDefinitionDataSource_basicByName
=== RUN   TestAccRoleDefinitionDataSource_builtIn_contributor
=== PAUSE TestAccRoleDefinitionDataSource_builtIn_contributor
=== RUN   TestAccRoleDefinitionDataSource_builtIn_owner
=== PAUSE TestAccRoleDefinitionDataSource_builtIn_owner
=== RUN   TestAccRoleDefinitionDataSource_builtIn_reader
=== PAUSE TestAccRoleDefinitionDataSource_builtIn_reader
=== RUN   TestAccRoleDefinitionDataSource_builtIn_virtualMachineContributor
=== PAUSE TestAccRoleDefinitionDataSource_builtIn_virtualMachineContributor
=== RUN   TestAccRoleDefinition_basic
=== PAUSE TestAccRoleDefinition_basic
=== RUN   TestAccRoleDefinition_requiresImport
=== PAUSE TestAccRoleDefinition_requiresImport
=== RUN   TestAccRoleDefinition_complete
=== PAUSE TestAccRoleDefinition_complete
=== RUN   TestAccRoleDefinition_update
=== PAUSE TestAccRoleDefinition_update
=== RUN   TestAccRoleDefinition_updateEmptyId
=== PAUSE TestAccRoleDefinition_updateEmptyId
=== RUN   TestAccRoleDefinition_emptyName
=== PAUSE TestAccRoleDefinition_emptyName
=== RUN   TestAccRoleDefinition_emptyPermissions
=== PAUSE TestAccRoleDefinition_emptyPermissions
=== RUN   TestAccRoleDefinition_managementGroup
=== PAUSE TestAccRoleDefinition_managementGroup
=== RUN   TestAccRoleDefinition_assignToSmallerScope
=== PAUSE TestAccRoleDefinition_assignToSmallerScope
=== RUN   TestAccRoleDefinition_noAssignableScope
=== PAUSE TestAccRoleDefinition_noAssignableScope
=== CONT  TestAccRoleDefinitionDataSource_basic
=== CONT  TestAccRoleDefinition_complete
=== CONT  TestAccRoleDefinition_emptyPermissions
=== CONT  TestAccRoleDefinition_requiresImport
=== CONT  TestAccRoleDefinition_assignToSmallerScope
=== CONT  TestAccRoleDefinition_noAssignableScope
=== CONT  TestAccRoleDefinitionDataSource_builtIn_reader
=== CONT  TestAccRoleDefinitionDataSource_basicByName
--- PASS: TestAccRoleDefinitionDataSource_builtIn_reader (20.41s)
=== CONT  TestAccRoleDefinition_managementGroup
    testcase.go:113: Step 1/2 error: Error running apply: exit status 1
        
        Error: unable to create Management Group "e271adc4-1e3b-4e42-a376-3cfafc1a2eba": managementgroups.Client#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="Permission to Microsoft.Management/managementGroups on resources of type 'Write' is required on the management group or its ancestors." Details=[{"raw":"Management Group ID: '/providers/Microsoft.Management/managementGroups/72f988bf-86f1-41af-91ab-2d7cd011db47'"}]
        
          with azurerm_management_group.test,
          on terraform_plugin_test.tf line 24, in resource "azurerm_management_group" "test":
          24: resource "azurerm_management_group" "test" {
        
--- FAIL: TestAccRoleDefinition_managementGroup (11.76s)
=== CONT  TestAccRoleDefinition_updateEmptyId
--- PASS: TestAccRoleDefinition_requiresImport (237.98s)
=== CONT  TestAccRoleDefinition_emptyName
--- PASS: TestAccRoleDefinition_emptyPermissions (323.85s)
=== CONT  TestAccRoleDefinition_basic
--- PASS: TestAccRoleDefinition_assignToSmallerScope (327.99s)
=== CONT  TestAccRoleDefinition_update
--- PASS: TestAccRoleDefinitionDataSource_basic (387.61s)
=== CONT  TestAccRoleDefinitionDataSource_builtIn_virtualMachineContributor
--- PASS: TestAccRoleDefinitionDataSource_builtIn_virtualMachineContributor (17.68s)
=== CONT  TestAccRoleDefinitionDataSource_builtIn_owner
--- PASS: TestAccRoleDefinitionDataSource_builtIn_owner (17.12s)
=== CONT  TestAccRoleDefinitionDataSource_builtIn_contributor
--- PASS: TestAccRoleDefinitionDataSource_builtIn_contributor (20.01s)
--- PASS: TestAccRoleDefinition_complete (454.16s)
--- PASS: TestAccRoleDefinition_emptyName (236.10s)
--- PASS: TestAccRoleDefinition_noAssignableScope (485.73s)
--- PASS: TestAccRoleDefinitionDataSource_basicByName (562.71s)
--- PASS: TestAccRoleDefinition_updateEmptyId (595.59s)
--- PASS: TestAccRoleDefinition_basic (321.54s)
--- PASS: TestAccRoleDefinition_update (673.80s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-azurerm/internal/services/authorization	1001.871s
FAIL
```
rerun the failed one on another subscription
```

terraform-provider-azurerm tengzh/rbac/sdk_swap*​ ≡
❯ tftest authorization TestAccRoleDefinition_managementGroup
=== RUN   TestAccRoleDefinition_managementGroup
=== PAUSE TestAccRoleDefinition_managementGroup
=== CONT  TestAccRoleDefinition_managementGroup
--- PASS: TestAccRoleDefinition_managementGroup (703.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/authorization	703.059s
```

`azurerm_role_assignment` also affected
image

### CODE REVIEW

File: internal/services/authorization/role_assignment_resource.go [hunk 1]
```diff
@@ -284,15 +291,24 @@ func resourceArmRoleAssignmentRead(d *pluginsdk.ResourceData, meta interface{})
 		d.Set("condition_version", props.ConditionVersion)
 
 		// allows for import when role name is used (also if the role name changes a plan will show a diff)
-		if roleId := props.RoleDefinitionID; roleId != nil {
-			roleResp, err := roleDefinitionsClient.GetByID(ctx, *roleId)
+		if roleDefResourceId := props.RoleDefinitionID; roleDefResourceId != nil {
+			// The role definition id returned does not contain scope when the scope is some special case.
+			// Such as management group. So we might need to add scope here.
```
Comment (magodo) [inline internal/services/authorization/role_assignment_resource.go:296]: Can we be more specific here that the modification is needed to workaround the issue: https://github.com/hashicorp/pandora/issues/3257 for `roledefinitions.ParseScopedRoleDefinitionID` for the roles assigned at the tenant level (both for tenant scope and mgmt group scope)?

**Review by magodo — CHANGES_REQUESTED** [review]

**Comment by ziyeqf** [comment]
I believe the CI failure is not caused by this PR..

**Comment by magodo** [comment]
LGTM!

File: internal/services/authorization/azuresdkhacks/definitions.go [hunk 2]
```diff
@@ -7,151 +7,119 @@ import (
...
-	Permissions *[]Permission `json:"permissions,omitempty"`
-	// AssignableScopes - Role definition assignable scopes.
-	AssignableScopes *[]string `json:"assignableScopes,omitempty"`
-
+	AssignableScopes *[]string                     `json:"assignableScopes,omitempty"`
+	Description      *string                       `json:"description,omitempty"`
+	Permissions      *[]roledefinitions.Permission `json:"permissions,omitempty"`
+	RoleName         *string                       `json:"roleName,omitempty"`
+	Type             *string                       `json:"type,omitempty"`
 	// not exposed in the sdk
```
Comment (tombuildsstuff) [inline internal/services/authorization/azuresdkhacks/definitions.go:120]: These fields are available [in the newer API version](https://github.com/hashicorp/go-azure-sdk/blob/5d2118ecfed532e1549552249c29398aea44653d/resource-manager/authorization/2022-04-01/roledefinitions/model_roledefinitionproperties.go#L14-L21) which is [already used by the Provider](https://github.com/hashicorp/terraform-provider-azurerm/blob/0d48f0d6f542d5a5b780e63b4e7d446660c69bfe/internal/services/authorization/client/client.go#L27) - as such I believe this workaround is no longer needed?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
Can we remove the workaround, since this is no longer needed?

File: internal/services/authorization/azuresdkhacks/definitions.go (hunk 2 above)
Comment (ziyeqf) [reply internal/services/authorization/azuresdkhacks/definitions.go:120]: I planed to remove them in another PR, and do the API version upgrade at that PR, how do you think?

File: internal/services/authorization/role_definition_data_source.go [hunk 3]
```diff
@@ -155,66 +155,82 @@ func (a RoleDefinitionDataSource) Read() sdk.ResourceFunc {
...
+				return fmt.Errorf("retirving %s: %+v", id, err)
+			}
+			if resp.Model == nil {
+				return fmt.Errorf("retirving %s: `Model` was nil", id)
+			}
+
+			role = *resp.Model
+
+			if role.Id == nil {
+				return fmt.Errorf("retirving %s: `Id` was nil", id)
```
Comment (mbfrahry) [inline internal/services/authorization/role_definition_data_source.go:205]:
```suggestion
				return fmt.Errorf("retrieving %s: `Id` was nil", id)
```

File: internal/services/authorization/role_definition_data_source.go [hunk 4]
```diff
@@ -155,66 +155,82 @@ func (a RoleDefinitionDataSource) Read() sdk.ResourceFunc {
...
-				}
+				id = roledefinitions.NewScopedRoleDefinitionID(config.Scope, defId)
+			}
+
+			resp, err := client.Get(ctx, id)
+			if err != nil {
+				return fmt.Errorf("retirving %s: %+v", id, err)
+			}
+			if resp.Model == nil {
+				return fmt.Errorf("retirving %s: `Model` was nil", id)
```
Comment (mbfrahry) [inline internal/services/authorization/role_definition_data_source.go:199]:
```suggestion
				return fmt.Errorf("retrieving %s: `Model` was nil", id)
```

File: internal/services/authorization/role_definition_data_source.go [hunk 5]
```diff
@@ -155,66 +155,82 @@ func (a RoleDefinitionDataSource) Read() sdk.ResourceFunc {
...
-				role, err = client.Get(ctx, config.Scope, defId)
-				if err != nil {
-					return fmt.Errorf("loading Role Definition: %+v", err)
-				}
+				id = roledefinitions.NewScopedRoleDefinitionID(config.Scope, defId)
+			}
+
+			resp, err := client.Get(ctx, id)
+			if err != nil {
+				return fmt.Errorf("retirving %s: %+v", id, err)
```
Comment (mbfrahry) [inline internal/services/authorization/role_definition_data_source.go:196]:
```suggestion
				return fmt.Errorf("retrieving %s: %+v", id, err)
```

**Review by mbfrahry — APPROVED** [review]
LGTM! thanks for this @ziyeqf

**Pull request merged**